### DISCUSSION TITLE
From pgvecto.rs to pgvector

### DISCUSSION THREAD

**Discussion by dani** [immich-app/immich#7296] [General]
Now that immich supports both pgvecto.rs and pgvector, I'd be interested in migrating to pgvector. I know I will miss some features, but pgvecto.rs is just a pain to manage (I'm not using the tensorchord Docker image as I run a centralized postgres server). Building the extension takes ages, Docker images are significantly bigger, upgrading from 0.1.11 to 0.2.0 seems not to work that well (need to debug this)
```
ERROR:  extension "vectors" has no update path from version "0.1.11" to version "0.2.0"
ERROR:  current transaction is aborted, commands ignored until end of transaction block
```
and the biggest issue : pgvecto.rs doesn't support logical replication, so prevent me from switching from one postgres instance to another (which I would otherwise do, as I run a patroni cluster). All in all, the advantages of pgvecto.rs do not seem to worth the maintainance cost.

Now, the biggest issue : it seems it's not supported to migrate from pgvecto.rs to pgvector. But is there some way to do it manually ?

**Comment by sovereignalmida** [comment]
I have the exact same issue with pgvector.rs right now. Immich wont run because of this error and I cannot make the upgrade of pgvector.rs work.

**Reply by dani** [reply]
Got it to work. The documentation to compile from source is not very explicit, but you have to copy the sql/upgrade/*.sql scripts to your postgres share/extension directory manually. You can see my Dockerfile for how to compile it : https://git.lapiole.org/nomad/common/src/branch/master/example/images/postgres/Dockerfile

Still, I think pgvecto.rs is not robust enough. It kind of work if you use the tensorchord Docker image, but not when you manage a real postgres server (with replication and PITR backups)

**Comment by sovereignalmida** [comment]
I got it to work too - I followed the steps here:

https://newreleases.io/project/github/immich-app/immich/release/v1.95.0

**Reply by dani** [reply]
Yes, but these steps are only if you use the tensorchord image. When building pgvecto.rs from source, it's a bit more complicated

**Comment by VoVAllen** [comment]
Hi, I'm the developer of pgvecto.rs. As an extension that constantly adds new features, unfortunately we need to make breaking changes periodically. So the upgrade process may be annoying at times, but we are also working hard to collaborate with immich's development team to reduce the pain of upgrading. In the latest version 0.2.1, we have introduced precompiled binary versions packaged in zip and deb format with the upgrade scripts, allowing non-docker users to install from it directly to avoid errors like `extension 'vectors' has no update path from version '0.1.11' to version '0.2.0'` when installing from source. 

And pgvecto.rs do support logical replication. Currently physical and streaming replication is not supported. Hope this helps!

**Reply by dani** [reply]
Yes my bad, I meant streaming replication (prevent from using patroni or other fancy stuff to manage postgres HA). Also, physical replication is most likely breaking my pgbackrest backups (haven't tested yet, but at best, I think if I try to restore I'll have either some empty, or worst, some corrupted tables...). Not using postgres internal storage system is the biggest issue for me. Breaking changes and recompilation are annoying, but I could live with it. Not being able to trust my backups, or do HA is another storry.

**Reply by VoVAllen** [reply]
Your data is safe with physical/streaming replication and also with pgbackrest. They're stored with postgres storage engine. Only the index part doesn't support physical/streaming replication. This means you need to manually recreate the index when using the standby database or recover from the pgbackrest.

**Reply by VoVAllen** [reply]
Also we understand such concerns from users, that's why we introduced the compatibility mode so that immich can support pgvector at the same time and user can choose based on their own tradeoff.

**Reply by dani** [reply]
OK, that's good to know. Still a pain to have to recreate the index (so not really possible to failover between nodes), but at least if the backups are safe...
Anyway, I appreciate the pgvector compatibility. Only problem is immich doesn't support switching to it with an existing instance

**Reply by evanrich** [reply]
> Hi, I'm the developer of pgvecto.rs. 
Any reason that the pgvecto.rs  dns no longer resolves?


![image](https://github.com/user-attachments/assets/800fed3b-9908-4c9e-ae02-7994004e1d61)

**Reply by WaterWei** [reply]
> > Hi, I'm the developer of pgvecto.rs.Hi, 我是 pgvecto.rs 的开发者。
> > Any reason that the pgvecto.rs  dns no longer resolves?pgvecto.rs 的 DNS 为何不再解析？
> 
> ![image](https://private-user-images.githubusercontent.com/518320/399380263-800fed3b-9908-4c9e-ae02-7994004e1d61.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY5MjI0NDIsIm5iZiI6MTczNjkyMjE0MiwicGF0aCI6Ii81MTgzMjAvMzk5MzgwMjYzLTgwMGZlZDNiLTk5MDgtNGM5ZS1hZTAyLTc5OTQwMDRlMWQ2MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDExNVQwNjIyMjJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZDYwZTUxMDZmNTBhYmMzZjFiMzM3OGEzOTliNjk1MTM5MzE1ODBiMWY0ZWYyYTE3YWE3ZWQ0Y2FmYmRkNWM0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7v33PtK93Rb5RcRObbh2mVQdT39hnQ5Pmtz98OTG8w4)

Because this domain expired ......  They migrated to the “vectorchord.ai“ .
